services/frontend-service/pkg/handler/release.go

```diff
@@ -17,35 +17,204 @@ Copyright 2023 freiheit.com*/
+				for _, sig := range form.File[fmt.Sprintf("signatures[%s]", environmentName)] {
+					if signature, err := readMultipartFile(sig); err != nil {
+						w.WriteHeader(500)
+						fmt.Fprintf(w, "Internal: %s", err)
+						return
+					} else {
+						if _, err := openpgp.CheckArmoredDetachedSignature(s.KeyRing, bytes.NewReader(content), bytes.NewReader(signature), nil); err != nil {
+							if err != pgperrors.ErrUnknownIssuer {
+								w.WriteHeader(400)
+								fmt.Fprintf(w, "Internal: Invalid Signature for %s: %s", k, err.Error())
+								return
+							}
+						} else {
+							validSignature = true
+							break
+						}
+					}
+				}
+				if !validSignature {
+					w.WriteHeader(400)
+					fmt.Fprintf(w, fmt.Sprintf("signature not found for %s", environmentName))
+					return
+				}
+
+			}
+
+			// TODO(HVG): validate that the manifest is valid yaml
+			tf.Manifests[environmentName] = string(content)
+		}
+
+	}
+	if len(tf.Manifests) == 0 {
+		w.WriteHeader(400)
+		fmt.Fprintf(w, "No manifest files provided")
+		return
+	}
+
+	if team, ok := form.Value["team"]; ok {
+		if len(team) == 1 {
+			tf.Team = team[0]
+		}
+	}
+
+	if sourceCommitId, ok := form.Value["source_commit_id"]; ok {
+		if len(sourceCommitId) == 1 && isCommitId(sourceCommitId[0]) {
+			tf.SourceCommitId = sourceCommitId[0]
+		}
 	}
-	auth.WriteUserToHttpHeader(req, *user)
 
-	//this is http in both!!
-	//	author gets lost because we are not writing to the context (I assume)
-	url, err := url.Parse(s.Config.HttpCdServer + "/release")
+	if sourceAuthor, ok := form.Value["source_author"]; ok {
+		if len(sourceAuthor) == 1 && isAuthor(sourceAuthor[0]) {
+			tf.SourceAuthor = sourceAuthor[0]
+		}
+	}
+
+	if sourceMessage, ok := form.Value["source_message"]; ok {
+		if len(sourceMessage) == 1 {
+			tf.SourceMessage = sourceMessage[0]
+		}
+	}
+	if version, ok := form.Value["version"]; ok {
+		if len(version) == 1 {
+			val, err := strconv.ParseUint(version[0], 10, 64)
+			if err != nil {
+				w.WriteHeader(400)
+				fmt.Fprintf(w, "Invalid version: %s", err)
+				return
+			}
+			tf.Version = val
+		}
+	}
+
+	response, err := s.BatchClient.ProcessBatch(ctx, &api.BatchRequest{Actions: []*api.BatchAction{
+		{
+			Action: &api.BatchAction_CreateRelease{
+				CreateRelease: &tf,
+			},
+		}},
+	})
 	if err != nil {
+		s, ok := status.FromError(err)
+		if ok && s.Code() == codes.InvalidArgument {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
 		http.Error(w, err.Error(), http.StatusInternalServerError)
 		return
 	}
-
-	// Release api requires file uploads in it, which is hard to implement in grpc.
-	// Release api in cd-service only exists in REST right now, so this endpoint directly calls the REST endpoint of cd-service instead of a grpc one.
-	cdServiceProxy := httputil.NewSingleHostReverseProxy(url)
-	cdServiceProxy.ServeHTTP(w, req)
+	jsonBlob, err := json.Marshal(response.Results[0].GetCreateReleaseResponse())
+	if err != nil {
+		return
```

Comment:
We should handle this somehow better. Just returning an empty body is not right thing.

Reply:
good point, I will return an internal server error and log an error.

Reply:
done